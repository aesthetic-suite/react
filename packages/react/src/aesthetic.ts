import { Aesthetic } from '@aesthetic/core';
import { createClientEngine } from '@aesthetic/style';

const aesthetic = new Aesthetic();

aesthetic.configureEngine(createClientEngine());

export default aesthetic;
