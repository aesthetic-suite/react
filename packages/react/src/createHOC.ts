/* eslint-disable @typescript-eslint/no-explicit-any */

import React from 'react';
import hoistNonReactStatics from 'hoist-non-react-statics';
import { WrapperComponent } from './types';

export default function createHOC<T extends React.FunctionComponent<any>>(
  name: string,
  WrappedComponent: React.ComponentType<any>,
  BaseComponent: T,
): T & WrapperComponent {
  const Component = BaseComponent as T & WrapperComponent;

  hoistNonReactStatics(Component, WrappedComponent);

  Component.displayName = `${name}(${WrappedComponent.displayName || WrappedComponent.name})`;
  Component.WrappedComponent = WrappedComponent;

  return Component;
}
