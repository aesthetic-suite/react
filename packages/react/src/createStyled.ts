import React from 'react';
import { createComponentStyles, LocalBlock, Utilities } from '@aesthetic/core';
import useStyles from './useStyles';

export default function createStyled<T extends keyof JSX.IntrinsicElements>(
  type: T,
  factory: (utilities: Utilities<LocalBlock>) => LocalBlock,
): React.ForwardRefExoticComponent<JSX.IntrinsicElements[T]> {
  if (__DEV__) {
    if (typeof factory !== 'function') {
      throw new TypeError(
        `Styled components require a style sheet factory function, found ${typeof factory}.`,
      );
    }
  }

  const styleSheet = createComponentStyles((utils) => ({
    element: factory(utils),
  }));

  const Component = React.forwardRef<unknown, JSX.IntrinsicElements[T]>((props, ref) => {
    const cx = useStyles(styleSheet);
    let className = cx('element');

    if (props.className) {
      className += ` ${props.className}`;
    }

    return React.createElement(type, {
      ...props,
      className,
      ref,
    });
  });

  Component.displayName = `styled(${type})`;

  return Component;
}
