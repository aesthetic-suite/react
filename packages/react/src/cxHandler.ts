import { ClassNameSheet, ClassNameSheetVariants, generateClassName } from '@aesthetic/core';
import { isObject, objectLoop } from '@aesthetic/utils';

const variantCache: Record<string, string[]> = {};

export default function cxHandler(keys: unknown[], classNames: ClassNameSheet<string>) {
  const names = keys as string[];
  let variantNames: string[] = [];

  if (isObject(names[0])) {
    const variants = (names.shift() as unknown) as ClassNameSheetVariants;
    const cacheKey = JSON.stringify(variants);
    const cache = variantCache[cacheKey];

    if (cache) {
      variantNames = cache;
    } else {
      objectLoop(variants, (value, type) => {
        variantNames.push(`${type}_${value}`);
      });

      variantCache[cacheKey] = variantNames;
    }
  }

  return generateClassName(names, variantNames, classNames);
}
