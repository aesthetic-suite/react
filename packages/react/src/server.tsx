import React from 'react';
import { formatDeclarationBlock, StyleEngine } from '@aesthetic/style';
import type { Sheet, SheetType } from '@aesthetic/types';

export * from '@aesthetic/style/server';

interface SheetProps {
  ruleIndex: number;
  sheet: Sheet;
  type: SheetType;
}

function Style({ ruleIndex, sheet, type }: SheetProps) {
  const lastIndex = sheet.cssRules.length - 1;

  // TODO switch to extractCssFromSheet when available
  let css = '';

  if (Object.keys(sheet.cssVariables).length > 0) {
    css += `:root { ${formatDeclarationBlock(sheet.cssVariables)} }`;
  }

  css += sheet.cssText;

  return (
    <style
      id={`aesthetic-${type}`}
      type="text/css"
      media="screen"
      data-aesthetic-type={type}
      data-aesthetic-hydrate-index={lastIndex}
      data-aesthetic-rule-index={ruleIndex}
    >
      {css}
    </style>
  );
}

export function renderToStyleElements(engine: StyleEngine): JSX.Element {
  return (
    <>
      <Style ruleIndex={engine.ruleIndex} sheet={engine.sheetManager.sheets.global} type="global" />

      <Style
        ruleIndex={engine.ruleIndex}
        sheet={engine.sheetManager.sheets.standard}
        type="standard"
      />

      <Style
        ruleIndex={engine.ruleIndex}
        sheet={engine.sheetManager.sheets.conditions}
        type="conditions"
      />
    </>
  );
}
