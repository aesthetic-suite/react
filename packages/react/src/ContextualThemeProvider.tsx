import React, { useState, useEffect } from 'react';
import aesthetic from './aesthetic';
import ThemeContext from './ThemeContext';
import useDirection from './useDirection';
import { ContextualThemeProviderProps } from './types';

/**
 * Provides a theme for a distinct portion of the React tree.
 */
export default function ContextualThemeProvider({
  children,
  name,
  wrapper,
}: ContextualThemeProviderProps) {
  const [className, setClassName] = useState('');
  const direction = useDirection();
  const theme = aesthetic.getTheme(name);

  // Render styles when theme/direction change
  useEffect(() => {
    setClassName(aesthetic.renderThemeStyles(theme, { direction }));
  }, [theme, direction]);

  return (
    <ThemeContext.Provider value={theme}>
      {React.cloneElement(wrapper, { className, 'data-theme': name }, children)}
    </ThemeContext.Provider>
  );
}
