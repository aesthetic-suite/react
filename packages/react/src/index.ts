/**
 * @copyright   2020, Miles Johnson
 * @license     https://opensource.org/licenses/MIT
 */

import { aesthetic } from './aesthetic';
import { createStyled } from './styles';

export * from './direction';
export * from './styles';
export * from './theme';
export * from './types';
export * from '@aesthetic/core';
export * from '@aesthetic/core-react';

export const {
	changeDirection,
	changeTheme,
	configure,
	createStyleSheet,
	createScopedStyleSheet,
	createThemeSheet,
	getActiveDirection,
	getActiveTheme,
	getEngine,
	getTheme,
	registerDefaultTheme,
	registerTheme,
	renderStyleSheet,
	renderFontFace,
	renderImport,
	renderKeyframes,
	renderThemeSheet,
	subscribe,
	unsubscribe,
} = aesthetic;

// Export aliases
export const style = createScopedStyleSheet;
export const styles = createStyleSheet;
export const styled = createStyled;
export const themeStyle = createThemeSheet;

// Do not use! This is for bundler pre-compilation usage!
// @private
export const internalAestheticRuntime = aesthetic;
