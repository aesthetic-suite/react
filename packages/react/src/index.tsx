/**
 * @copyright   2020, Miles Johnson
 * @license     https://opensource.org/licenses/MIT
 */

import aesthetic from './aesthetic';
import ThemeContext from './ThemeContext';
import ThemeProvider from './ThemeProvider';
import ContextualThemeProvider from './ContextualThemeProvider';
import createStyled from './createStyled';
import useStyles from './useStyles';
import useTheme from './useTheme';
import withStyles from './withStyles';
import withTheme from './withTheme';

export * from '@aesthetic/core';
export * from './direction';
export * from './types';

export {
  ThemeContext,
  ThemeProvider,
  ContextualThemeProvider,
  createStyled,
  useStyles,
  useTheme,
  withStyles,
  withTheme,
};

export const {
  changeDirection,
  changeTheme,
  configure,
  createComponentStyles,
  createThemeStyles,
  generateClassName,
  getActiveDirection,
  getActiveTheme,
  getEngine,
  getTheme,
  registerDefaultTheme,
  registerTheme,
  renderComponentStyles,
  renderFontFace,
  renderImport,
  renderKeyframes,
  renderThemeStyles,
  subscribe,
  unsubscribe,
} = aesthetic;
