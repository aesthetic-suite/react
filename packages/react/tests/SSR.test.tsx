/* eslint-disable react/jsx-no-literals */

import React from 'react';
import { renderToStaticMarkup, renderToString } from 'react-dom/server';
import { ServerStyleEngine } from '@aesthetic/style';
import { createServerEngine, renderToStyleMarkup } from '@aesthetic/style/server';
import { ThemeProvider, useStyles } from '../src';
import { ButtonProps, createStyleSheet } from './__mocks__/Button';
import { setupAestheticReact, teardownAestheticReact } from './helpers';

describe('SSR', () => {
  let engine: ServerStyleEngine;

  function Button({ children, block, disabled, large, small }: ButtonProps) {
    const cx = useStyles(createStyleSheet());

    return (
      <button
        type="button"
        className={cx(
          {
            // eslint-disable-next-line no-nested-ternary
            size: large ? 'lg' : small ? 'sm' : 'df',
          },
          'button',
          block && 'button_block',
          disabled && 'button_disabled',
        )}
      >
        {children}
      </button>
    );
  }

  function App() {
    return (
      <ThemeProvider name="twilight">
        <main>
          <div>You are not logged in!</div>
          <Button href="/login">Login</Button>
          <Button href="/register">Register</Button>
        </main>
      </ThemeProvider>
    );
  }

  beforeEach(() => {
    setupAestheticReact();

    engine = createServerEngine();
  });

  afterEach(() => {
    teardownAestheticReact();

    // @ts-expect-error
    delete global.AESTHETIC_CUSTOM_ENGINE;
  });

  describe('renderToString()', () => {
    it('renders markup', () => {
      expect(renderToString(engine.extractStyles(<App />))).toMatchSnapshot();
      expect(renderToStyleMarkup(engine)).toMatchSnapshot();
    });
  });

  describe('renderToStaticMarkup()', () => {
    it('renders markup', () => {
      expect(renderToStaticMarkup(engine.extractStyles(<App />))).toMatchSnapshot();
      expect(renderToStyleMarkup(engine)).toMatchSnapshot();
    });
  });
});
