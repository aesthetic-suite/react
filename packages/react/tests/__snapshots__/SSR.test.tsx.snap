// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SSR renderToStaticMarkup() renders markup 1`] = `"<main><div>You are not logged in!</div><button type=\\"button\\" class=\\"a b c d e f g h i j k l m n o p q r s t u v w z a1\\">Login</button><button type=\\"button\\" class=\\"a b c d e f g h i j k l m n o p q r s t u v w z a1\\">Register</button></main>"`;

exports[`SSR renderToStaticMarkup() renders markup 2`] = `"<style  data-aesthetic-hydrate-index=\\"-1\\" data-aesthetic-rule-index=\\"40\\" data-aesthetic-type=\\"global\\" id=\\"aesthetic-global\\" media=\\"screen\\" type=\\"text/css\\"></style><style  data-aesthetic-hydrate-index=\\"40\\" data-aesthetic-rule-index=\\"40\\" data-aesthetic-type=\\"standard\\" id=\\"aesthetic-standard\\" media=\\"screen\\" type=\\"text/css\\">.a { appearance:none; }.b { background-color:var(--palette-brand-color-40); }.c { border:0; }.d { cursor:pointer; }.e { display:inline-flex; }.f { font-size:inherit; }.g { margin:0; }.h { padding:0; }.i { text-decoration:none; }.j { user-select:auto; }.k { vertical-align:middle; }.l { position:relative; }.m { text-align:center; }.n { color:var(--palette-neutral-color-00); }.o { border-color:var(--palette-neutral-color-40); }.p { border-style:solid; }.q { border-width:var(--border-df-width); }.r { border-radius:var(--border-df-radius); }.s:hover { background-color:var(--palette-brand-color-50); }.t:active { background-color:var(--palette-brand-color-60); }.u::-moz-focus-inner { border:0; }.v::-moz-focus-inner { padding:0; }.w::-moz-focus-inner { margin:0; }.x { min-width:7.50rem; }.y { padding:var(--spacing-sm) var(--spacing-df); }.z { min-width:10rem; }.a1 { padding:var(--spacing-df) var(--spacing-lg); }.b1 { min-width:12.50rem; }.c1 { padding:var(--spacing-lg) var(--spacing-xl); }.d1 { transition:all 1s; }.e1 { overflow-wrap:normal; }.f1 { word-wrap:normal; }.g1 { word-break:normal; }.h1 { display:block; }.i1 { width:100%; }.j1 { white-space:normal; }.k1 { overflow:hidden; }.l1 { opacity:0.4; }.m1 { opacity:0.5; }.n1 { opacity:0.6; }.o1 { transition:none; }</style><style  data-aesthetic-hydrate-index=\\"-1\\" data-aesthetic-rule-index=\\"40\\" data-aesthetic-type=\\"conditions\\" id=\\"aesthetic-conditions\\" media=\\"screen\\" type=\\"text/css\\"></style>"`;

exports[`SSR renderToString() renders markup 1`] = `"<main><div>You are not logged in!</div><button type=\\"button\\" class=\\"a b c d e f g h i j k l m n o p q r s t u v w z a1\\">Login</button><button type=\\"button\\" class=\\"a b c d e f g h i j k l m n o p q r s t u v w z a1\\">Register</button></main>"`;

exports[`SSR renderToString() renders markup 2`] = `"<style  data-aesthetic-hydrate-index=\\"-1\\" data-aesthetic-rule-index=\\"40\\" data-aesthetic-type=\\"global\\" id=\\"aesthetic-global\\" media=\\"screen\\" type=\\"text/css\\"></style><style  data-aesthetic-hydrate-index=\\"40\\" data-aesthetic-rule-index=\\"40\\" data-aesthetic-type=\\"standard\\" id=\\"aesthetic-standard\\" media=\\"screen\\" type=\\"text/css\\">.a { appearance:none; }.b { background-color:var(--palette-brand-color-40); }.c { border:0; }.d { cursor:pointer; }.e { display:inline-flex; }.f { font-size:inherit; }.g { margin:0; }.h { padding:0; }.i { text-decoration:none; }.j { user-select:auto; }.k { vertical-align:middle; }.l { position:relative; }.m { text-align:center; }.n { color:var(--palette-neutral-color-00); }.o { border-color:var(--palette-neutral-color-40); }.p { border-style:solid; }.q { border-width:var(--border-df-width); }.r { border-radius:var(--border-df-radius); }.s:hover { background-color:var(--palette-brand-color-50); }.t:active { background-color:var(--palette-brand-color-60); }.u::-moz-focus-inner { border:0; }.v::-moz-focus-inner { padding:0; }.w::-moz-focus-inner { margin:0; }.x { min-width:7.50rem; }.y { padding:var(--spacing-sm) var(--spacing-df); }.z { min-width:10rem; }.a1 { padding:var(--spacing-df) var(--spacing-lg); }.b1 { min-width:12.50rem; }.c1 { padding:var(--spacing-lg) var(--spacing-xl); }.d1 { transition:all 1s; }.e1 { overflow-wrap:normal; }.f1 { word-wrap:normal; }.g1 { word-break:normal; }.h1 { display:block; }.i1 { width:100%; }.j1 { white-space:normal; }.k1 { overflow:hidden; }.l1 { opacity:0.4; }.m1 { opacity:0.5; }.n1 { opacity:0.6; }.o1 { transition:none; }</style><style  data-aesthetic-hydrate-index=\\"-1\\" data-aesthetic-rule-index=\\"40\\" data-aesthetic-type=\\"conditions\\" id=\\"aesthetic-conditions\\" media=\\"screen\\" type=\\"text/css\\"></style>"`;
